### PR TITLE
graphcast improvements part II

### PR DESCRIPTION
**Pull request by mnabian**
# Modulus Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: examples/weather/graphcast/conf/config.yaml [hunk 1]
```diff
@@ -16,58 +16,121 @@
...
+                            # Apex is pre-installed in Modulus containers.
+
+# ┌───────────────────────────────────────────┐
+# │           Dataset Configuration           │
+# └───────────────────────────────────────────┘
+
+dataset_path: /data/era5_75var             # Path to the dataset.
+static_dataset_path: static                # Path to the static datasets. Includes .nc files for land-sea mask and geopotential.
+dataset_metadata_path: /data/era5_75var/metadata/data.json  # Path to the dataset metadata, containing channel names.
+time_diff_std_path: /code/mnabian/graphcast/PR/pr/modulus_next/examples/weather/graphcast/time_diff_std_new.npy  # Path to the .npy file with standard deviation of normalized per-variable per-pressure level time differences.
```
Comment (dallasfoster) [inline examples/weather/graphcast/conf/config.yaml:70]: Maybe reconsider what these paths are set as in this public config file.

File: modulus/utils/graphcast/loss.py [hunk 2]
```diff
@@ -110,3 +113,116 @@ def forward(self, invar: torch.Tensor, outvar: torch.Tensor) -> torch.Tensor:
...
+        """
+
+        # outvar normalization
+        loss = (invar - outvar) ** 2  # [T,C,H,W]
+        # weighted by inverse variance
+        loss = (
+            loss
+            * 1.0
+            / torch.square(self.time_diff_str.view(1, -1, 1, 1).to(loss.device))
+        )
```
Comment (dallasfoster) [inline modulus/utils/graphcast/loss.py:152]: Going back to the `time_diff` script, I strongly believe that the diffs should be weighted by area if that is also done here.

**Review by dallasfoster — COMMENTED** [review]
Configs look really good, just a few small comments.

File: examples/weather/graphcast/compute_time_diff_std.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    mean, mean_sqr = 0, 0
+    for i, data in enumerate(datapipe):
+        invar = data[0]["invar"]
+        outvar = data[0]["outvar"][0]
+        diff = outvar - invar
+        diff_sqr = torch.square(diff)
+
+        mean += torch.mean(diff, dim=(2, 3)) / len(datapipe)
+        mean_sqr += torch.mean(diff_sqr, dim=(2, 3)) / len(datapipe)
```
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/weather/graphcast/compute_time_diff_std.py]: Thanks. Added area weighting

File: examples/weather/graphcast/conf/config.yaml (hunk 1 above)
Comment (mnabian) [reply examples/weather/graphcast/conf/config.yaml]: Fixed it

File: modulus/utils/graphcast/loss.py (hunk 2 above)
Comment (mnabian) [reply modulus/utils/graphcast/loss.py:152]: Added area weighting

**Review by dallasfoster — APPROVED** [review]
Looks good.

File: examples/weather/graphcast/compute_time_diff_std.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        num_workers=cfg.num_workers,
+        device=dist.device,
+        process_rank=dist.rank,
+        world_size=dist.world_size,
+        shuffle=False,
+    )
+    print(f"Loaded training datapipe of length {len(datapipe)}")
+
+    area = (
+        grid_cell_area(torch.linspace(-90, 90, steps=cfg.latlon_res[0]), unit="deg")
```
Comment (stadlmax) [inline examples/weather/graphcast/compute_time_diff_std.py:56]: is this always correct or in-sync with the model definition? can we maybe add an utility that defines these lat-lon meshes? in the training, the "area" is computed from the `model.latitude` values directly

side-note: the area function expects the inputs to be in a -90/90 range anyways, a 0/180 range e.g. would require to replace the cos-term with a sin-term.

File: examples/weather/graphcast/conf/config.yaml [hunk 5]
```diff
@@ -16,58 +16,121 @@
...
+cugraphops_decoder: false   # If true, uses cugraphops backend for the decoder.
+recompute_activation: true  # If true, recomputes activation in backward to save memory. Currently, only SiLU is supported.
+use_apex: true              # If true, uses Apex for fused Adam optimizer, typically resulting in 10-20% faster training iterations.
+                            # Apex is pre-installed in Modulus containers.
+
+# ┌───────────────────────────────────────────┐
+# │           Dataset Configuration           │
+# └───────────────────────────────────────────┘
+
+dataset_path: /data/era5_75var             # Path to the dataset.
```
Comment (stadlmax) [inline examples/weather/graphcast/conf/config.yaml:67]: Are these relativ or absolute paths?

**Review by stadlmax — CHANGES_REQUESTED** [review]
Looks good overall, a few minor comments here and there

File: examples/weather/graphcast/conf/config.yaml (hunk 5 above)
Comment (mnabian) [reply examples/weather/graphcast/conf/config.yaml:67]: These are absolute paths

File: examples/weather/graphcast/compute_time_diff_std.py (hunk 4 above)
Comment (mnabian) [reply examples/weather/graphcast/compute_time_diff_std.py:56]: Yes, this is in sync with the model definition. Although I think both are inconsistent with the ERA5 grid, which goes from 90 to -90. However, because cell area is proportional to cosine of latitudes, this does not change the final values.

File: examples/weather/graphcast/README.md [hunk 6]
```diff
@@ -25,8 +28,26 @@ to transfer data.
...
+icosahedron, incorporates edges of varying scales,
+enabling the model to effectively learn atmospheric processes across diverse spatial
+extents from large-scale
+pressure systems to localized wind patterns. Finally, the decoder translates the processed
+information into a future weather forecast.
+This GNN-based approach deviates from conventional grid-structured models, offering a more
+flexible framework for modeling intricate
+interactions between geographically separated atmospheric variables.
+
+![GraphCast model schematic. Image is taken from the GraphCast paper.](../../../docs/img/bloodflow_1d_mgn_geometries.png)
```
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/weather/graphcast/README.md:47]: Oh thanks for this!

File: modulus/utils/graphcast/loss.py [hunk 7]
```diff
@@ -110,3 +113,116 @@ def forward(self, invar: torch.Tensor, outvar: torch.Tensor) -> torch.Tensor:
...
+        # weighted by variables
+        variable_weights = self.variable_weights.view(1, -1, 1, 1).to(loss.device)
+        loss = loss * variable_weights  # [T,C,H,W]
+        # weighted by area
+        loss = loss.mean(dim=(0, 1))
+        loss = torch.mul(loss, self.area)
+        loss = loss.mean()
+        return loss
+
+    def get_time_diff_str(self, time_diff_std_path, channels_list):
```
[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/utils/graphcast/loss.py:162]: Fixed it

File: examples/weather/graphcast/train_graphcast.py [hunk 8]
```diff
@@ -192,16 +204,16 @@ def __init__(self, cfg: DictConfig, dist, rank_zero_logger):
         self.model.train()
 
         # get area
-        if hasattr(self.model, "module"):
-            self.area = grid_cell_area(
-                self.model.module.lat_lon_grid[:, :, 0], unit="deg"
-            )
-        else:
-            self.area = grid_cell_area(self.model.lat_lon_grid[:, :, 0], unit="deg")
+        self.area = grid_cell_area(self.lat_lon_grid[:, :, 0], unit="deg")
```
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/weather/graphcast/train_graphcast.py:207]: Thanks for your comment. I agree that utilities like this should be part of the main code base. NVR is working on an earth2-grid package which will be used by Modulus as a dependency, and I think general utilities like this will end up there. For now I just renamed the utility as you suggested. Your understanding on the normalized area is correct, and that was my thinking as well.

File: examples/weather/graphcast/compute_time_diff_std.py (hunk 4 above)
Comment (stadlmax) [reply examples/weather/graphcast/compute_time_diff_std.py]: The main issue arises from this being inconsistent with the model, i.e. if you would have different area weights due to slightly different definitions.

Comment (stadlmax) [reply examples/weather/graphcast/compute_time_diff_std.py]: I guess for now, this should be fine though.

File: modulus/utils/graphcast/loss.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (stadlmax) [reply modulus/utils/graphcast/loss.py:162]: Don't want to be too nitpicky, but sure that you changed this? :)
Or is the "str" actually intentional?

[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/utils/graphcast/loss.py]: Oh sorry maybe I forgot to push the change. It is fixed now

**Review by dallasfoster — APPROVED** [review]

**Review by stadlmax — APPROVED** [review]
LGTM
